### PR TITLE
net/bluetooth: Add LE create connection

### PR DESCRIPTION
**Pull request by amitpurwar**
Signed-off-by: Amit Purwar <amit.purwar@samsung.com>

### CODE REVIEW

**Comment by seinfra** [comment]
Target : [85b1ceed1d8d5222b513e1346ea83b67d18cc2a8] - Code Rule Check (C++) OK.

**Comment by seinfra** [comment]
Target : [85b1ceed1d8d5222b513e1346ea83b67d18cc2a8] - Code Rule Check Result:
os/net/bluetooth/bluetooth.c:1119: ERROR: [IDT_M_SCH] switch and case should be at the same indent

**Comment by seinfra** [comment]
Target : [fa65474cc104554dc96c7aad38261ee65172000d] - Code Rule Check (C++) OK.

**Comment by seinfra** [comment]
Target : [fa65474cc104554dc96c7aad38261ee65172000d] - Code Rule Check OK.

File: os/net/bluetooth/bt_conn.c [hunk 1]
```diff
@@ -718,6 +835,15 @@ void bt_conn_release(FAR struct bt_conn_s *conn)
 	bt_addr_le_copy(&conn->dst, BT_ADDR_LE_ANY);
 }
 
+struct bt_conn_s * bt_conn_relref(FAR struct bt_conn_s *conn)
```
Comment (dh79pyun) [inline os/net/bluetooth/bt_conn.c]: bt_conn_relref is the original function name? or bt_conn_release_ref is a better name for this function

Comment (amitpurwar) [reply os/net/bluetooth/bt_conn.c]: bt_conn_relref in not original name. I will modify it to bt_conn_release_ref .

**Comment by seinfra** [comment]
Target : [ee41054de38243471cca2fc593bc79b17cbba508] - Code Rule Check (C++) OK.

**Comment by seinfra** [comment]
Target : [ee41054de38243471cca2fc593bc79b17cbba508] - Code Rule Check Result:
os/net/bluetooth/bluetooth.c:1119: ERROR: [IDT_M_SCH] switch and case should be at the same indent

**Comment by seinfra** [comment]
Target : [ccea93ba01b026990bd9ecb195bc9db489d0e6b7] - Code Rule Check (C++) OK.

**Comment by seinfra** [comment]
Target : [ccea93ba01b026990bd9ecb195bc9db489d0e6b7] - Code Rule Check OK.

**Comment by seinfra** [comment]
Target : [04597babc36fc750b33a3a8ef4e7dd175fdce642] - Code Rule Check (C++) OK.

**Comment by seinfra** [comment]
Target : [04597babc36fc750b33a3a8ef4e7dd175fdce642] - Code Rule Check OK.

File: os/net/bluetooth/bluetooth.c [hunk 2]
```diff
@@ -1325,8 +1461,11 @@ static int start_le_scan(uint8_t scan_type, uint16_t interval, uint16_t window)
 	return 0;
 }
 
-int bt_ble_scan_update(bool fast_scan)
+int bt_le_scan_update_internal(bool fast_scan)
```
Comment (Wootak) [inline os/net/bluetooth/bluetooth.c]: Please add 'static' keyword.

File: os/net/bluetooth/bluetooth.c [hunk 3]
```diff
@@ -1084,19 +1092,147 @@ int bt_conn_le_param_update(struct bt_conn *conn, const struct bt_le_conn_param
...
+	bt_conn_set_param_le(conn, param);
+
+	bt_conn_set_state(conn, BT_CONN_CONNECT_SCAN);
+
+	bt_le_scan_update_internal(true);
+
+	return (struct bt_conn *)conn;
 }
 
 int bt_le_set_auto_conn(const bt_addr_le_t *addr, const struct bt_le_conn_param *param)
```
Comment (Wootak) [inline os/net/bluetooth/bluetooth.c]: Please add 'static' keyword.

File: os/net/bluetooth/bt_conn.h [hunk 4]
```diff
@@ -406,4 +449,21 @@ int bt_conn_le_start_encryption(FAR struct bt_conn_s *conn, uint64_t rand, uint1
...
 
+int bt_conn_addr_le_cmp(const struct bt_conn_s *conn, const bt_addr_le_t *peer);
+
+struct bt_conn_s *bt_conn_lookup_state_le(const bt_addr_le_t *peer, const enum bt_conn_state_e state);
+
+struct bt_conn_s *bt_conn_lookup_addr_le_id(uint8_t id, const bt_addr_le_t *peer);
+
+void bt_conn_set_param_le(struct bt_conn_s *conn, const struct bt_le_conn_param *param);
+
+struct bt_conn_s *bt_conn_add_le(const bt_addr_le_t *peer);
```
Comment (Wootak) [inline os/net/bluetooth/bt_conn.h]: Please remove this definition in conn_internal.h

File: os/net/bluetooth/bt_conn.h [hunk 5]
```diff
@@ -406,4 +449,21 @@ int bt_conn_le_start_encryption(FAR struct bt_conn_s *conn, uint64_t rand, uint1
 
 int bt_conn_le_conn_update(FAR struct bt_conn_s *conn, uint16_t min, uint16_t max, uint16_t latency, uint16_t timeout);
 
+int bt_conn_addr_le_cmp(const struct bt_conn_s *conn, const bt_addr_le_t *peer);
+
+struct bt_conn_s *bt_conn_lookup_state_le(const bt_addr_le_t *peer, const enum bt_conn_state_e state);
```
Comment (Wootak) [inline os/net/bluetooth/bt_conn.h]: Please remove this definition in conn_internal.h.

File: os/net/bluetooth/bluetooth.c [hunk 6]
```diff
@@ -1042,20 +1042,28 @@ void bt_foreach_bond(uint8_t id, void (*func)(const struct bt_bond_info *info, v
...
 	return;
 }
 
 struct bt_conn *bt_conn_lookup_addr_le(uint8_t id, const bt_addr_le_t *peer)
 {
-	// TODO: need to implement
-	return NULL;
+	struct bt_conn_s * conns;
+
+	conns = bt_conn_lookup_addr_le_id(id, peer);
```
Comment (Wootak) [inline os/net/bluetooth/bluetooth.c]: I think we don't need to add this new function.
we can do that in here.

File: os/net/bluetooth/bt_conn.c [hunk 7]
```diff
@@ -626,6 +677,30 @@ FAR struct bt_conn_s *bt_conn_lookup_addr_le_internal(FAR const bt_addr_le_t *pe
 	return NULL;
 }
 
+struct bt_conn_s *bt_conn_lookup_addr_le_id(uint8_t id, const bt_addr_le_t *peer)
```
Comment (Wootak) [inline os/net/bluetooth/bt_conn.c]: Same as above, please move this functionality in 'bt_conn_lookup_addr_le'.

File: os/net/bluetooth/bluetooth.c (hunk 2 above)
Comment (amitpurwar) [reply os/net/bluetooth/bluetooth.c]: Made the bt_le_scan_update_internal function static

**Comment by seinfra** [comment]
Target : [e1e2b31eeda53651f65f2c063e76618be61e44b3] - Code Rule Check (C++) OK.

**Comment by seinfra** [comment]
Target : [e1e2b31eeda53651f65f2c063e76618be61e44b3] - Code Rule Check OK.

File: os/net/bluetooth/bluetooth.c (hunk 6 above)
Comment (amitpurwar) [reply os/net/bluetooth/bluetooth.c]: We are using bt_conn_lookup_addr_le_id() at some other places also.

File: os/net/bluetooth/bt_conn.c (hunk 7 above)
Comment (amitpurwar) [reply os/net/bluetooth/bt_conn.c]: We are using bt_conn_lookup_addr_le_id() at some other places also.

File: os/net/bluetooth/bt_conn.h (hunk 5 above)
Comment (amitpurwar) [reply os/net/bluetooth/bt_conn.h]: Removed bt_conn_lookup_state_le from conn_internal.h.

File: os/net/bluetooth/bt_conn.h (hunk 4 above)
Comment (amitpurwar) [reply os/net/bluetooth/bt_conn.h]: Removed bt_conn_add_le from conn_internal.h.

File: os/net/bluetooth/bluetooth.c (hunk 3 above)
Comment (amitpurwar) [reply os/net/bluetooth/bluetooth.c]: bt_le_set_auto_conn() is defined in include/tinyara/bluetooth/conn.h
So it could not be static

**Comment by seinfra** [comment]
Target : [5f0ce090b2b6345e759d9d3a46f7addf0a2f6db2] - Code Rule Check OK.

**Comment by seinfra** [comment]
Target : [5f0ce090b2b6345e759d9d3a46f7addf0a2f6db2] - Code Rule Check (C++) OK.

**Comment by seinfra** [comment]
Target : [432110a3134f50577930431b757ff1c0d03f0999] - Code Rule Check (C++) OK.

**Comment by seinfra** [comment]
Target : [432110a3134f50577930431b757ff1c0d03f0999] - Code Rule Check OK.

**Pull request merged**